scripts/esbuild/util.ts

```diff
@@ -0,0 +1,124 @@
+export function getEsbuildAliases(): Record<string, string> {
+  return {
+    // node module redirection
+    chalk: 'ansi-colors',
+
+    // mapping aliases to top-level bundle entrypoints
+    '@stencil/core/testing': './testing/index.js',
+    '@stencil/core/compiler': './compiler/stencil.js',
+    '@stencil/core/dev-server': './dev-server/index.js',
+    '@stencil/core/mock-doc': './mock-doc/index.cjs',
+    '@stencil/core/internal/testing': './internal/testing/index.js',
+
+    // mapping node.js module names to `sys/node` "cache"
+    //
+    // these allow us to bundle and ship a dependency (like `prompts`) as part
+    // of the Stencil distributable but also have our separate bundles
+    // reference the same file
+    prompts: './sys/node/prompts.js',
+  };
+}
+
+/**
+ * Node modules which should be universally marked as external
+ *
+ * Note that we should not rely on this to mark node.js built-in modules as
+ * external. Doing so will override esbuild's automatic marking of those modules
+ * as side-effect-free, which allows imports from them to be properly
+ * tree-shaken.
+ */
+const externalNodeModules = [
+  '@jest/core',
+  '@jest/reporters',
+  '@microsoft/typescript-etw',
+  'assert',
+  'buffer',
+  'child_process',
+  'console',
+  'constants',
+  'expect',
+  'fsevents',
+  'inspector',
+  'jest',
+  'jest-cli',
+  'jest-config',
+  'jest-message-id',
+  'jest-pnp-resolver',
+  'jest-runner',
+  'net',
+  'puppeteer',
+  'puppeteer-core',
+  'readline',
+  'stream',
+  'tty',
+  'url',
+  'util',
+  'vm',
+  'yargs',
+  'zlib',
+];
+
+/**
+ * Get a manifest of modules which should be marked as external for a given
+ * esbuild bundle
+ *
+ * @param opts options for the current build
+ * @param ownEntryPoint the entry point alias of the current module
+ * @returns a list of modules which should be marked as external
+ */
+export function getEsbuildExternalModules(opts: BuildOptions, ownEntryPoint: string): string[] {
+  const bundles = Object.values(opts.output);
+  const externalBundles = bundles.filter((outdir) => outdir !== ownEntryPoint).map((outdir) => join(outdir, '*'));
+
+  return [...externalNodeModules, ...externalBundles];
+}
+
+/**
+ * A helper which runs an array of esbuild, uh, _builds_
+ *
+ * This accepts an array of build configurations and will either run a synchronous
+ * build _or_ run them all in watch mode, depending on the {@link BuildOptions['isWatch']}
+ * setting.
+ *
+ * @param builds the array of outputs to build
+ * @param opts Stencil build options
+ * @returns a Promise representing the execution of the builds
+ */
+export function runBuilds(builds: ESBuildOptions[], opts: BuildOptions): Promise<(void | ESBuildResult)[]> {
+  if (opts.isWatch) {
+    return Promise.all(
+      builds.map(async (buildConfig) => {
+        const context = await esbuild.context(buildConfig);
+        return context.watch();
+      }),
+    );
+  } else {
+    return Promise.all(builds.map(esbuild.build));
+  }
+}
```

Comment:
this lil helper will take an array of esbuild configs and our own Stencil-specific `BuildOptions` object and if the `isWatch` property is set on the `BuildOptions` it will run a watch-mode build using [`esbuild.context`](https://esbuild.github.io/api/#watch), otherwise it will just use `esbuild.build` to do a single build.

The idea is that each of our bundle functions would create a series of esbuild configurations, do any other setup that is necessary, and then return the result of calling this function with those build configs. Then in our main entry point we use `Promise.all` to await the result of all the builds.

also kind of funny that the esbuild interface is _also_ called `BuildOptions`, so I've taken to aliasing it as `ESBuildOptions`